### ISSUE TITLE
`Op::Unary(arg, UnaryOp::Silu)` branch in Tensor::backward() may cause NaN

### ISSUE THREAD

**Issue by Adamska1008** [huggingface/candle#2082]
I tried to fine-tune tinyllama model while I implemented the model structure and the Lora layer independently. During the train process the loss went to NaN, and after 2 thousand years I finally found that the first NaN occurs in the backward process of Silu。Here is related code
```rust
Op::Unary(arg, UnaryOp::Silu) => {
    let sum_grad = grads.or_insert(arg)?;
    // d/dx silu = sigmoid(x) * (1 + x * (1 - sigmoid(x)))
    let mut sigmoid_arg = (*node / arg)?;
    let silu_grad = (&sigmoid_arg * (1. + (arg * (1. - &sigmoid_arg)?)?)?)?;
    *sum_grad = sum_grad.add(&(&grad * silu_grad)?)?;
}
```
`(*node / arg)` here may cause div-zero and result in NaN. I fix it by adding eps, is there any better way to fix it?

**Comment by LaurentMazare** [comment]
Ah thanks for reporting this, that's indeed an issue, I just made #2083 that should fix it in a slightly less efficient way (the tricky bit with a epsilon is that it could still return some nans or some very unsable values if you're unlucky).

**Comment by Adamska1008** [comment]
> Ah thanks for reporting this, that's indeed an issue, I just made #2083 that should fix it in a slightly less efficient way (the tricky bit with a epsilon is that it could still return some nans or some very unsable values if you're unlucky).

Great, I switch to git dependency in cargo.toml and it works well.

**Comment by singhaki** [comment]
@Adamska1008 can your share lora finetuning code? or guide in that direction?

**Comment by Adamska1008** [comment]
> @Adamska1008 can your share lora finetuning code? or guide in that direction?

you can try this candle-lora here https://github.com/EricLBuehler/candle-lora
